File: pexman2016_calgary/generate_prompts.R
# =============================================================================
# generate_prompts.R
# -----------------------------------------------------------------------------
# Adapted from the guenther2023ViSpa pipeline 
# Calgary Semantic Decision Project (Pexman et al., 2016)
#
# Reads processed/exp1.csv and writes one JSON prompt per participant to prompts.jsonl.zip
#
# =============================================================================

library(tidyverse)
library(jsonlite)

# Resolve paths from this script's location so it runs from any working
# directory and always writes inside its own study folder.
.args <- commandArgs(trailingOnly = FALSE)
SCRIPT_DIR <- dirname(sub("^--file=", "", .args[grep("^--file=", .args)]))
if (length(SCRIPT_DIR) == 0 || !nzchar(SCRIPT_DIR)) SCRIPT_DIR <- getwd()


# (removed setwd(rstudioapi::...) -- it cannot work outside RStudio; see SCRIPT_DIR above)

# -----------------------------------------------------------------------------
# INSTRUCTION
# -----------------------------------------------------------------------------

EXPERIMENT_NAME <- "pexman2016calgary"

# Instructions from paper
INSTRUCTION <- "In this task, you will see a series of words presented one at a time. For each word, you need to decide whether it refers to something concrete or abstract.
Concrete words are defined as things or actions in reality, which you can experience directly through your senses. These words are experience-based. Night, bridle, and lynx are examples of concrete words.
Abstract words are defined as something you cannot experience directly through your senses or actions. These words are language-based, as their meaning depends on other words. Have, limitation, and outspokenness are examples of abstract words.
Note that words are not restricted to nouns — verbs and adjectives can also be concrete or abstract."

# Feedback for incorrect responses and timeouts
TRIAL_CORRECT   <- "Trial %d: The word is '%s'. You press <<%s>>.\n"
TRIAL_INCORRECT <- "Trial %d: The word is '%s'. You press <<%s>>. Incorrect.\n"
TRIAL_TIMEOUT   <- "Trial %d: The word is '%s'. No response detected.\n"

# -----------------------------------------------------------------------------
# Helpers
# -----------------------------------------------------------------------------

# Build JSON manually so separators match Python's json.dumps output
# String escaping handled by jsonlite
format_jsonl_line <- function(text, experiment, participant_id, rt, age, gender) {
  paste0(
    '{"text": ',            toJSON(text,           auto_unbox = TRUE),
    ', "experiment": ',     toJSON(experiment,     auto_unbox = TRUE),
    ', "participant_id": ', toJSON(participant_id, auto_unbox = TRUE),
    ', "rt": ',             toJSON(as.integer(rt), auto_unbox = FALSE, na = "null"),
    ', "age": ',            toJSON(as.integer(age), auto_unbox = TRUE, na = "null"),
    ', "gender": ',         toJSON(as.character(gender), auto_unbox = TRUE, na = "null"),
    '}'
  )
}

# Build the full prompt for one participant: instruction + replay of trials.
build_participant_prompt <- function(df_participant, trial_indices) {
  
  # Dominant hand for this participant
  dominant_hand <- df_participant$ehi_class[1]
  
  if (dominant_hand == "R") {
    concrete_hand <- "right"
    abstract_hand <- "left"
  } else if (dominant_hand == "L") {
    concrete_hand <- "left"
    abstract_hand <- "right"
  } else {
    # Ambidextrous (so preferred writing hand)
    concrete_hand <- "dominant"
    abstract_hand <- "non-dominant"
  }
  # Add instruction based on handedness
  instruction <- paste0(
    INSTRUCTION,
    sprintf("Press the %s button for concrete words and the %s button for abstract words.\n\n",
            concrete_hand, abstract_hand)
  )
  
  prompt <- instruction
  for (trial in trial_indices) {
    row <- df_participant[df_participant$trial_id == trial, , drop = FALSE]
    if (nrow(row) > 0) {
      if (is.na(row$response) || row$response == "T") {
        prompt <- paste0(prompt, sprintf(TRIAL_TIMEOUT, trial + 1, row$stimulus)) # Trial info + feedback if timeout or no response 
      } else if (row$accuracy == 1L) {
        response_hand <- ifelse(row$response == "C", concrete_hand, abstract_hand)
        prompt <- paste0(prompt, sprintf(TRIAL_CORRECT, trial + 1, row$stimulus, response_hand)) # Trial info + no feedback if response correct
      } else {
        response_hand <- ifelse(row$response == "C", concrete_hand, abstract_hand)
        prompt <- paste0(prompt, sprintf(TRIAL_INCORRECT, trial + 1, row$stimulus, response_hand)) # Trial info + feedback if not correct
      }
    }
  }  
  paste0(prompt, "\n")
}

# -----------------------------------------------------------------------------
# Main
# -----------------------------------------------------------------------------

main <- function() {
  df <- read_csv(file.path(SCRIPT_DIR, "processed_data", "exp1.csv"), show_col_types = FALSE)

  df <- df[order(df$participant_id, df$trial_order), , drop = FALSE]
  df$participant_id <- match(df$participant_id, unique(df$participant_id))
  
  participants <- unique(df$participant_id)
  trial_indices <- 0:max(df$trial_id)
  
  con <- file(file.path(SCRIPT_DIR, "prompts.jsonl"), open = "wb")
  
  for (p in participants) {
    df_p   <- df[df$participant_id == p, , drop = FALSE]
    prompt <- build_participant_prompt(df_p, trial_indices)
    line   <- format_jsonl_line(prompt, EXPERIMENT_NAME, p,
                                df_p$rt, df_p$age[1], df_p$gender[1])
    writeLines(line, con, sep = "\n")
  }
  
  close(con) # manually because it omitted the last participant when done like in guenther script
  zip(file.path(SCRIPT_DIR, "prompts.jsonl.zip"), file.path(SCRIPT_DIR, "prompts.jsonl"), flags = "-j9X")
  file.remove(file.path(SCRIPT_DIR, "prompts.jsonl"))
  cat("Done. Written", length(participants), "prompts to prompts.jsonl.zip\n")
}

if (sys.nframe() == 0) {
  main()
}

File: hsieh2025_rating/generate_prompts.py
# Generating prompts for hsieh2025_rating

import pandas as pd
import jsonlines

from pathlib import Path

# Resolve paths from this script's location so the script runs from any
# working directory, on any machine.
SCRIPT_DIR = Path(__file__).resolve().parent

# Load data
df = pd.read_csv(SCRIPT_DIR / "processed_data" / "exp1.csv")

# create empty list to store all prompts
all_prompts = []

# Instructions

instruction = "此份問卷每道題目皆為由兩個中文字所構成的詞語，其中大多數的詞彙您可能從未見過。請您根據直覺判斷能否想到該詞語可能的意義，若認為「幾乎無法」，請填答1；若認為「非常容易」，請填答5。以下有三個範例僅供參考（無需填答），其中有些可能對您來說很容易聯想到某個意思，有些可能很難，不過並沒有標準答案，您可以憑自身語感試著賦予這些詞彙意義。 "

# Generate individual prompts for participants
participant_list = df['participant_id'].unique()
trial_list = df['trial_id'].unique()

for participant in participant_list:
    exp_participant = df[df['participant_id'] == participant]
    #age = exp_participant['age'].iloc[0].item()
    individual_prompt = instruction
    for trial in trial_list:
        exp_trial = exp_participant.loc[exp_participant['trial_id'] == trial]
        if not exp_trial.empty:  # Only process if trial exists for this participant
            stimulus = exp_trial['stimulus'].iloc[0]
            response = exp_trial['response'].iloc[0]
            datapoint = f'試驗{trial}：詞語為{stimulus}。你填答<<{response}>>。\n'
            individual_prompt += datapoint
    all_prompts.append({'text': individual_prompt, 'experiment': 'hsieh2025_rating/dataA', 'participant_id': int(participant)})

# Save all prompts to JSONL file
with jsonlines.open(SCRIPT_DIR / "prompts.jsonl", "w") as writer:
    writer.write_all(all_prompts)
